File: bpe.rs
use std::collections::HashMap;
use std::env;
use std::fs::File;
use std::io::{Read, Write};
use std::cmp::Ordering;

fn main() -> std::io::Result<()> {
    let mut data: Vec<u16> = Vec::new();
    let fname = env::args().nth(1).unwrap_or_else(|| "enw3".to_string());

    load_data(&fname, &mut data)?;

    let mut next_value: u16 = 256;
    let mut thesaurus: HashMap<u16, Vec<u8>> = (0..=255).map(|x| (x as u16, vec![x as u8])).collect();

    let mut ph = calc_pair_histo(&mut data);

    for i in 1..=65000 {
        if one_step(i, &mut data, &mut ph, &mut next_value, &mut thesaurus) {
            break;
        }
        if i % 1000 == 0 {
            if let Err(e) = save_tokens(&fname, &data) {
                eprintln!("{}", e);
            }
        }
    }
    save_tokens(&fname, &data)?;

    Ok(())
}

fn load_data(fname: &str, data: &mut Vec<u16>) -> std::io::Result<()> {
    let mut file = File::open(fname)?;
    let mut buffer = [0u8; 2000000];
    loop {
        let bytes_read = file.read(&mut buffer)?;
        if bytes_read == 0 {
            break;
        }
        data.extend(buffer[..bytes_read].iter().map(|&x| x as u16));
    }
    Ok(())
}

struct WalkIter {
    pos: usize,
}

impl WalkIter {
    fn new(data: &Vec<u16>) -> Self {
        let mut iter = WalkIter { pos: 0 };
        iter.skip_holes(data);
        iter
    }

    fn skip_holes(&mut self, data: &Vec<u16>) {
        while self.pos < data.len() && data[self.pos] == 0 {
            self.pos += 1;
        }
    }

    fn next(&mut self, data: &Vec<u16>) -> Option<usize> {
        if self.pos >= data.len() {
            None
        } else {
            let p = self.pos;
            self.pos += 1;
            self.skip_holes(data);
            Some(p)
        }
    }
}

struct PairWalker {
    walker: WalkIter,
    p1: Option<usize>,
    p2: Option<usize>,
}

impl PairWalker {
    fn new(data: &Vec<u16>) -> Self {
        let mut walker = WalkIter::new(data);
        let p1 = walker.next(data);
        let p2 = walker.next(data);
        PairWalker { walker, p1, p2 }
    }

    fn next(&mut self, data: &Vec<u16>) -> Option<(usize, usize)> {
        if let (Some(a), Some(b)) = (self.p1, self.p2) {
            self.p1 = self.p2;
            self.p2 = self.walker.next(data);
            Some((a, b))
        } else {
            None
        }
    }

    fn after_replace(&mut self, replace_pos: usize, data: &Vec<u16>) -> (Option<usize>, Option<usize>) {
        self.p1 = self.p2;
        self.p2 = self.walker.next(data);
        let prev_idx = if replace_pos > 0 {
            let mut i = replace_pos - 1;
            while i > 0 && data[i] == 0 {
                i -= 1;
            }
            if data[i] != 0 { Some(i) } else { None }
        } else { None };
        (prev_idx, self.p1)
    }
}

#[derive(Hash, Eq, PartialEq, Clone, Copy, Debug)]
struct VPair {
    value1: u16,
    value2: u16,
}

impl Ord for VPair {
    fn cmp(&self, other: &Self) -> Ordering {
        self.value1.cmp(&other.value1).then(self.value2.cmp(&other.value2))
    }
}

impl PartialOrd for VPair {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn calc_pair_histo(data: &Vec<u16>) -> HashMap<VPair, usize> {
    let mut h = HashMap::new();
    let mut pairs = PairWalker::new(data);
    while let Some(p) = pairs.next(data) {
        *h.entry(VPair { value1: data[p.0], value2: data[p.1] }).or_insert(0) += 1;
    }
    h
}

fn most_freq_vals<K: Ord + Copy>(h: &HashMap<K, usize>) -> (K, usize, usize) {
    let mut min_top = None;
    let mut max_n = 0;
    let mut total = 0;
    for (&k, &n) in h.iter() {
        total += n;
        if n > max_n {
            max_n = n;
            min_top = Some(k);
        } else if n == max_n {
            min_top = Some(min_top.map_or(k, |mt| mt.min(k)));
        }
    }
    (min_top.unwrap(), max_n, total)
}

fn replace_pair(data: &mut Vec<u16>, old_value_pair: VPair, new_val: u16, ph: &mut HashMap<VPair, usize>) {
    let mut pairs = PairWalker::new(data);
    while let Some(index_pair) = pairs.next(data) {
        let vp = VPair { value1: data[index_pair.0], value2: data[index_pair.1] };
        if vp == old_value_pair {
            data[index_pair.0] = new_val;
            data[index_pair.1] = 0;
            let (prev_idx_opt, next_idx_opt) = pairs.after_replace(index_pair.0, data);
            if let Some(prev_idx) = prev_idx_opt {
                let left_val = data[prev_idx];
                *ph.get_mut(&VPair { value1: left_val, value2: vp.value1 }).unwrap() -= 1;
                *ph.entry(VPair { value1: left_val, value2: new_val }).or_insert(0) += 1;
            }
            if let Some(next_idx) = next_idx_opt {
                let right_val = data[next_idx];
                *ph.get_mut(&VPair { value1: vp.value2, value2: right_val }).unwrap() -= 1;
                *ph.entry(VPair { value1: new_val, value2: right_val }).or_insert(0) += 1;
            }
        }
    }
}

fn one_step(step: usize, data: &mut Vec<u16>, ph: &mut HashMap<VPair, usize>, next_value: &mut u16, thesaurus: &mut HashMap<u16, Vec<u8>>) -> bool {
    let (vp, n, total) = most_freq_vals(ph);
    if n == 1 {
        return true;
    }
    if (total as f64) < (data.len() as f64) * 0.707 {
        *data = data.iter().filter(|&&x| x != 0).cloned().collect();
    }
    if step % 100 == 0 {
        println!("Step {}: n={} {:?} -> {}", step, n, vp, next_value);
    }
    replace_pair(data, vp, *next_value, ph);
    *ph.get_mut(&vp).unwrap() = 0;
    let mut new_entry = thesaurus[&vp.value1].clone();
    new_entry.extend_from_slice(&thesaurus[&vp.value2]);
    thesaurus.insert(*next_value, new_entry);
    *next_value += 1;
    false
}

fn save_tokens(fname: &str, data: &[u16]) -> std::io::Result<()> {
    let chunk_size = 1000000;
    let file_path = format!("{}.rtok", fname);
    let mut file = File::create(&file_path)?;

    let mut buffer = Vec::with_capacity(chunk_size);
    let mut h = HashMap::new();

    for &value in data.iter().filter(|&&x| x != 0) {
        buffer.push(value);
        *h.entry(value).or_insert(0) += 1;
        if buffer.len() >= chunk_size {
            file.write_all(unsafe { std::slice::from_raw_parts(buffer.as_ptr() as *const u8, buffer.len() * 2) })?;
            buffer.clear();
        }
    }

    if !buffer.is_empty() {
        file.write_all(unsafe { std::slice::from_raw_parts(buffer.as_ptr() as *const u8, buffer.len() * 2) })?;
    }

    Ok(())
}

File: bpe.js
const fs = require('fs');

let data = [];
const fname = process.argv.length > 2 ? process.argv[2] : "enw3";

function loadData() {
    return new Promise((resolve, reject) => {
        const readStream = fs.createReadStream(fname);
        readStream.on('data', (chunk) => {
            for (let x of chunk) {
                data.push(x);
            }
        });
        readStream.on('end', resolve);
        readStream.on('error', reject);
    });
}

class WalkIter {
    constructor() {
        this.pos = 0;
        this.skipHoles();
    }

    skipHoles() {
        while (this.pos < data.length && data[this.pos] === 0) {
            this.pos++;
        }
    }

    next() {
        if (this.pos >= data.length) return null;
        const p = this.pos;
        this.pos++;
        this.skipHoles();
        return p;
    }
}

class PairWalker {
    constructor() {
        this.walker = new WalkIter();
        this.p1 = this.walker.next();
        this.p2 = this.walker.next();
    }

    next() {
        if (this.p1 !== null && this.p2 !== null) {
            const result = [this.p1, this.p2];
            this.p1 = this.p2;
            this.p2 = this.walker.next();
            return result;
        } else {
            return null;
        }
    }

    afterReplace(replacePos) {
        this.p1 = this.p2;
        this.p2 = this.walker.next();
        let prevIdx = null;
        if (replacePos > 0) {
            let i = replacePos - 1;
            while (data[i] === 0 && i > 0) i--;
            if (data[i] !== 0) prevIdx = i;
        }
        return [prevIdx, this.p1];
    }
}

class VPair {
    constructor(value1, value2) {
        this.value1 = value1;
        this.value2 = value2;
    }

    toString() {
        return `${this.value1},${this.value2}`;
    }

    static fromString(str) {
        const [value1, value2] = str.split(',').map(Number);
        return new VPair(value1, value2);
    }

    lessThan(b) {
        if (this.value1 !== b.value1) {
            return this.value1 < b.value1;
        }
        return this.value2 < b.value2;
    }

}

function calcPairHisto() {
    const h = new Map();
    const pairs = new PairWalker();
    let p;
    while ((p = pairs.next()) !== null) {
        const key = new VPair(data[p[0]], data[p[1]]).toString();
        h.set(key, (h.get(key) || 0) + 1);
    }
    return h;
}

function mostFreqVals(h) {
    let minTop = null;
    let maxN = 0;
    let total = 0;
    for (const [k, n] of h.entries()) {
        total += n;
        if (n > maxN) {
            maxN = n;
            minTop = k;
        } else if (n === maxN && n > 0) {
            let kpair = VPair.fromString(k);
            let minTopPair = VPair.fromString(minTop);
            minTop = kpair.lessThan(minTopPair) ? k : minTop;
        }
    }
    return [VPair.fromString(minTop), maxN, total];
}

function replacePair(oldValuePair, newVal, ph) {
    const pairs = new PairWalker();
    let indexPair;
    while ((indexPair = pairs.next()) !== null) {
        const vp = new VPair(data[indexPair[0]], data[indexPair[1]]);
        if (vp.value1 === oldValuePair.value1 && vp.value2 === oldValuePair.value2) {
            data[indexPair[0]] = newVal;
            data[indexPair[1]] = 0;
            const [prevIdxOpt, nextIdxOpt] = pairs.afterReplace(indexPair[0]);
            if (prevIdxOpt !== null) {
                const leftVal = data[prevIdxOpt];
                ph.set(`${leftVal},${vp.value1}`, ph.get(`${leftVal},${vp.value1}`) - 1);
                const newLeftPair = `${leftVal},${newVal}`;
                ph.set(newLeftPair, (ph.get(newLeftPair) || 0) + 1);
            }
            if (nextIdxOpt !== null) {
                const rightVal = data[nextIdxOpt];
                ph.set(`${vp.value2},${rightVal}`, ph.get(`${vp.value2},${rightVal}`) - 1);
                const newRightPair = `${newVal},${rightVal}`;
                ph.set(newRightPair, (ph.get(newRightPair) || 0) + 1);
            }
        }
    }
}

let nextValue = 256;
const thesaurus = new Map();

function oneStep(step, ph) {
    const [vp, n, total] = mostFreqVals(ph);
    if (n === 1) return true;
    if (total < data.length * 0.707) {
        const compacted = [];
        const iter = new WalkIter();
        let pos;
        while ((pos = iter.next()) !== null) {
            compacted.push(data[pos]);
        }
        data = compacted;
    }
    if (step % 100 === 0) {
        console.log(`Step ${step}: n=${n} ${vp.toString()} -> ${nextValue}`);
    }
    replacePair(vp, nextValue, ph);
    ph.set(vp.toString(), 0);
    thesaurus.set(nextValue, [...thesaurus.get(vp.value1), ...thesaurus.get(vp.value2)]);
    nextValue++;
    return false;
}

function saveTokens() {
    const chunkSize = 1000000;
    const filePath = `${fname}.jstok`;
    const writeStream = fs.createWriteStream(filePath);
    const buffer = Buffer.alloc(chunkSize * 2);
    let bufferPos = 0;
    const w = new WalkIter();
    const h = new Map();

    let pos;
    while ((pos = w.next()) !== null) {
        buffer.writeUInt16LE(data[pos], bufferPos);
        bufferPos += 2;
        h.set(data[pos], (h.get(data[pos]) || 0) + 1);
        if (bufferPos >= chunkSize * 2) {
            writeStream.write(buffer.slice(0, bufferPos));
            bufferPos = 0;
        }
    }

    if (bufferPos > 0) {
        writeStream.write(buffer.slice(0, bufferPos));
    }

    writeStream.end();
}

async function main() {
    for (let x = 0; x <= 255; x++) {
        thesaurus.set(x, [x]);
    }

    await loadData();
    let ph = calcPairHisto();

    for (let i = 1; i <= 65000; i++) {
        if (oneStep(i, ph)) break;
        if (i % 1000 === 0) {
            try {
                await saveTokens();
            } catch (error) {
                console.error(error);
            }
        }
    }
    await saveTokens();
}

main().catch(console.error);
